swarm/test/cluster.go

```diff
@@ -0,0 +1,228 @@
+	cluster.Nodes = make([]*Node, 0, size)
+	for i := 0; i < size; i++ {
+		dir := filepath.Join(cluster.TmpDir, fmt.Sprintf("swarm%02d", i))
+		if err := os.Mkdir(dir, 0700); err != nil {
+			cluster.Shutdown()
+			return nil, err
+		}
+
+		node, err := NewNode(dir)
+		if err != nil {
+			cluster.Shutdown()
+			return nil, err
+		}
+		node.Name = fmt.Sprintf("swarm%02d", i)
+
+		cluster.Nodes = append(cluster.Nodes, node)
+	}
+
+	if size == 1 {
+		return cluster, nil
+	}
+
+	// connect the nodes together
+	for _, node := range cluster.Nodes {
+		if err := node.Client.Call(nil, "admin_addPeer", cluster.Nodes[0].Enode); err != nil {
+			cluster.Shutdown()
+			return nil, err
+		}
+	}
+
+	// wait until all nodes have the correct number of peers
+outer:
+	for _, node := range cluster.Nodes {
+		var peers []*p2p.PeerInfo
+		for start := time.Now(); time.Since(start) < 10*time.Second; time.Sleep(50 * time.Millisecond) {
+			if err := node.Client.Call(&peers, "admin_peers"); err != nil {
+				cluster.Shutdown()
+				return nil, err
+			}
+			if len(peers) == len(cluster.Nodes)-1 {
+				continue outer
+			}
+		}
+		return nil, fmt.Errorf("%s only has %d / %d peers", node.Name, len(peers), len(cluster.Nodes)-1)
+	}
+
+	return cluster, nil
+}
+
+func MustCluster(size int) *Cluster {
+	cluster, err := NewCluster(size)
+	if err != nil {
+		panic(err)
+	}
+	return cluster
+}
+
+func (c *Cluster) Shutdown() {
+	for _, node := range c.Nodes {
+		node.Shutdown()
+	}
+	os.RemoveAll(c.TmpDir)
+}
+
+type Node struct {
+	Name    string
+	Addr    string
+	URL     string
+	Enode   string
+	Dir     string
+	Client  *rpc.Client
+	Process *os.Process
+	LogFile *os.File
+}
+
+const Passphrase = "swarm-test-passphrase"
+
+func NewNode(dir string) (*Node, error) {
+	// create key
+	conf := &node.Config{
+		DataDir: dir,
+	}
+	n, err := node.New(conf)
+	if err != nil {
+		return nil, err
+	}
+	account, err := n.AccountManager().Backends(keystore.KeyStoreType)[0].(*keystore.KeyStore).NewAccount(Passphrase)
+	if err != nil {
+		return nil, err
+	}
+
+	node := &Node{Dir: dir}
+
+	// create the log file
+	logFile, err := os.Create(filepath.Join(dir, "swarm.log"))
+	if err != nil {
+		return nil, err
+	}
+	node.LogFile = logFile
+
+	// assign ports
+	httpPort, err := assignTCPPort()
+	if err != nil {
+		return nil, err
+	}
+	p2pPort, err := assignTCPPort()
+	if err != nil {
+		return nil, err
+	}
+
+	// start the node
+	cmd := exec.Command(
+		"swarm",
+		"--port", p2pPort,
+		"--nodiscover",
+		"--datadir", dir,
+		"--ethapi", "",
+		"--bzzaccount", account.Address.String(),
+		"--bzznetworkid", "321",
+		"--bzzport", httpPort,
+		"--verbosity", "6",
+	)
+	cmd.Stdin = strings.NewReader(Passphrase)
+	cmd.Stdout = logFile
+	cmd.Stderr = logFile
+	if err := cmd.Start(); err != nil {
+		node.Shutdown()
+		return nil, err
+	}
+	node.Process = cmd.Process
+
+	// wait for the node to start
+	for start := time.Now(); time.Since(start) < 10*time.Second; time.Sleep(50 * time.Millisecond) {
```

Comment:
There is a non trivial amount of time between exec'ing the process and it listening on the IPC socket, a retry dial loop seems to be the easiest way to wait for that to happen.